src/bin/cargo/commands/add.rs

```diff
@@ -242,7 +242,22 @@ fn parse_dependencies(config: &Config, matches: &ArgMatches) -> CargoResult<Vec<
         .flatten()
         .map(|c| (Some(c.clone()), None))
         .collect::<IndexMap<_, _>>();
+
     let mut infer_crate_name = false;
+
+    for (crate_name, _) in crates.iter() {
+        let crate_name = crate_name.as_ref().unwrap();
+        if !crate_name.starts_with("+") {
+            continue;
+        }
+
+        let crate_name_wo_prefix = &crate_name[1..];
+        anyhow::bail!(
+            "invalid package name: \"{crate_name}\"
+    Use `cargo {crate_name} add` if you meant to use the `{crate_name_wo_prefix}` toolchain."
+        );
```

Comment:
Just did a second pass.

* Probably could simplify to this?
* _“invalid character \`{ch}\` in package name: \`{pkg}\`"_ is already [a thing in Cargo](https://github.com/rust-lang/cargo/blob/1548f3bec4d5fcb9152c347966bf4228b76aaeba/src/cargo/util/restricted_names.rs#L60). Feel like we can follow that pattern. And it clearly points out the bad part.
* In Cargo error messages we tend to use backtick to quote stuff.
* and ditto for `cargo install`

```suggestion
        if let Some (toolchain) = crate_name.strip_prefix('+') {
            anyhow::bail!(
                "invalid character `+` in package name: `+{toolchain}`
    Use `cargo +{toolchain} add` if you meant to use the `{toolchain}` toolchain."
            );
        }
        
```

Reply:
Implemented in the new commits.